cip/1.accepted/CIP2017-02-07-Map-Projection.adoc

```diff
@@ -0,0 +1,121 @@
+= CIP2017-02-07 Map Projection
+:numbered:
+:toc:
+:toc-placement: macro
+:source-highlighter: codemirror
+
+*Author:* Tobias Lindaaker <tobias.lindaaker@neotechnology.com>
+
+toc::[]
+
+== Map Projection
+
+Map projection takes a map or entity (node or relationship) and creates a map based on selected properties from map or entity.
+The map projection largely reuses the map syntax of curly braces and comma separated entries.
+But the projection entries are just a single identifier starting with a dot (.).
+The projected map entries use the same key in the projected map as in the source map.
+
+The following example:
+
+[source, cypher]
+----
+someMap {.alpha, .beta}
+----
+
+is thus equivalent to:
+
+[source, cypher]
+----
+{alpha: someMap.alpha, beta: someMap.beta}
+----
+
+In order to be further useful, map projection allow other types of selectors to augment the projected map.
+
+Projected maps are also allowed as a `ReturnItem` in the `RETURN` and `WITH` clauses, in this case the projected map occurs without a preceding variable and instead projects variables from the current scope into a map.
+While both _property selectors_ and _variable selectors_ are syntactically allowed in this case, the semantics of both of these selectors are equivalent.
+
+=== Selectors
+
+There are four different types of entries allowed in a map projection:
+
+• *Property selector* - includes a property from the original map into the projected map using the same key in the projected map as in the original map.
+  This selector optionally also allows further projection of a nested map, if the selected property is a map or entity.
+• *Variable selector* - includes an entry into the projected map based on a variable in the current scope.
+  The key for the entry in the projected map will be the same as the name of the variable.
+  This selector optionally also allows further projection of a nested map, if the selected variable is a map or entity.
+• *Literal entry* - includes a entry into the projected map using the regular map entry syntax, i.e. with the value being explicitly different from the key.
+  In this case the value can be any arbitrary expression, just like in a regular map literal.
+• *All properties selector* - includes all properties from the original map or entity into the projected map. +
+  The all properties selector also allows marking properties for exclusion that would otherwise have been included by the selector. +
```

Comment:
Yes, that is intentional. The second sentence is specific point about all properties selectors. I thought it was worth highlighting that, so I put it on its own line, since I thought a single sub-bullet was too much weight for it.